File: scenario_service.py
"""
Scenario Service - Historical scenario backtesting for portfolios

Uses yfinance to fetch actual historical data during specific market events
and calculates how a given portfolio would have performed.

This is a NEW service and does NOT modify or depend on dataload/ files.
"""

import yfinance as yf
import pandas as pd
import numpy as np
from datetime import datetime

class ScenarioService:
    """Calculate portfolio performance during historical scenarios"""
    
    # Historical scenario date ranges
    SCENARIOS = {
        '2008-financial-crisis': {
            'start': '2007-10-09',
            'end': '2009-03-09',
            'name': '2008 Financial Crisis'
        },
        'covid-2020-crash': {
            'start': '2020-02-19',
            'end': '2020-03-23',
            'name': 'COVID-19 Crash'
        },
        'covid-recovery-2020': {
            'start': '2020-03-23',
            'end': '2021-11-08',
            'name': 'COVID Recovery'
        },
        'dot-com-bubble': {
            'start': '2000-03-10',
            'end': '2002-10-09',
            'name': 'Dot-Com Bubble'
        },
        'post-2008-recovery': {
            'start': '2009-03-09',
            'end': '2020-02-19',
            'name': 'Post-2008 Bull Market'
        },
        'crypto-winter-2022': {
            'start': '2021-11-10',
            'end': '2022-11-21',
            'name': '2022 Crypto Winter'
        },
        'black-monday-1987': {
            'start': '1987-10-14',
            'end': '1987-10-19',
            'name': 'Black Monday 1987'
        },
        '2022-bear-market': {
            'start': '2022-01-03',
            'end': '2022-10-12',
            'name': '2022 Bear Market'
        }
    }
    
    # Proxy tickers for asset classes when specific ticker isn't available
    PROXY_TICKERS = {
        'stock': '^GSPC',    # S&P 500 index with deep history
        'etf': '^GSPC',      # ETF fallback to market index for older periods
        'bond': '^TNX',      # 10Y Treasury yield proxy with older history
        'crypto': 'BTC-USD', # Bitcoin (available from ~2014)
        'commodity': 'GC=F'  # Gold futures with older history
    }

    CRYPTO_SYMBOL_MAP = {
        'BTC': 'BTC-USD',
        'ETH': 'ETH-USD',
        'SOL': 'SOL-USD',
        'ADA': 'ADA-USD',
        'DOT': 'DOT-USD',
        'BNB': 'BNB-USD',
        'XRP': 'XRP-USD',
        'DOGE': 'DOGE-USD',
    }

    KNOWN_INCEPTION_DATES = {
        'SPY': '1993-01-29',
        'QQQ': '1999-03-10',
        'NVDA': '1999-01-22',
        'TSLA': '2010-06-29',
        'BTC-USD': '2014-09-17',
        'ETH-USD': '2017-11-09',
        'SOL-USD': '2020-04-10',
        'ADA-USD': '2017-10-01',
        'DOT-USD': '2020-08-20',
        'BNB-USD': '2017-10-01',
        'XRP-USD': '2014-09-17',
        'DOGE-USD': '2014-09-17',
    }
    
    def __init__(self):
        self._price_cache = {}

    def _is_before_symbol_inception(self, symbol, start_date, end_date):
        inception = self.KNOWN_INCEPTION_DATES.get(symbol)
        if not inception:
            return False

        period_end = datetime.strptime(end_date, '%Y-%m-%d').date()
        inception_date = datetime.strptime(inception, '%Y-%m-%d').date()
        return period_end < inception_date

    def _normalize_symbol(self, symbol, asset_class):
        s = str(symbol).strip().upper()

        if asset_class == 'crypto':
            if s.endswith('-USD'):
                return s
            return self.CRYPTO_SYMBOL_MAP.get(s, f"{s}-USD")

        return s

    def _extract_close_series(self, data):
        if data is None or data.empty:
            return None

        close_data = data.get('Close')
        if close_data is None:
            return None

        if isinstance(close_data, pd.DataFrame):
            # yfinance may return a single-column DataFrame or MultiIndex columns.
            close_series = close_data.iloc[:, 0]
        else:
            close_series = close_data

        close_series = pd.to_numeric(close_series, errors='coerce').dropna()
        if close_series.empty:
            return None

        return close_series

    def _download_close_series(self, symbol, start_date, end_date):
        cache_key = (symbol, start_date, end_date)
        if cache_key in self._price_cache:
            return self._price_cache[cache_key]

        if self._is_before_symbol_inception(symbol, start_date, end_date):
            self._price_cache[cache_key] = None
            return None

        data = yf.download(
            symbol,
            start=start_date,
            end=end_date,
            progress=False,
            auto_adjust=True,
            threads=False,
        )
        close_series = self._extract_close_series(data)
        self._price_cache[cache_key] = close_series
        return close_series

    def _get_effective_series(self, symbol, asset_class, start_date, end_date):
        normalized_symbol = self._normalize_symbol(symbol, asset_class)
        symbol_series = self._download_close_series(normalized_symbol, start_date, end_date)

        if symbol_series is not None and len(symbol_series) >= 2:
            return symbol_series, normalized_symbol, False

        proxy = self.PROXY_TICKERS.get(asset_class, 'SPY')
        proxy_series = self._download_close_series(proxy, start_date, end_date)

        if proxy_series is not None and len(proxy_series) >= 2:
            return proxy_series, proxy, True

        return None, normalized_symbol, True
    
    def calculate_scenario(self, holdings, scenario_id):
        """
        Calculate how portfolio would perform in historical scenario
        
        Args:
            holdings: List of dicts with {symbol, quantity, currentValue, assetClass}
            scenario_id: ID of the scenario to simulate
            
        Returns:
            Dict with totalReturn, maxDrawdown, volatility, affectedAssets, timeline
        """
        if scenario_id not in self.SCENARIOS:
            raise ValueError(f"Unknown scenario: {scenario_id}")
        
        scenario = self.SCENARIOS[scenario_id]
        start_date = scenario['start']
        end_date = scenario['end']

        # Reset request-scoped cache to keep repeated downloads low and deterministic.
        self._price_cache = {}
        
        # Calculate initial portfolio value
        initial_value = float(sum(float(h.get('currentValue', 0) or 0) for h in holdings))
        if initial_value <= 0:
            raise ValueError("Portfolio total value must be greater than zero")
        
        # Fetch historical data for all holdings
        affected_assets = []
        for holding in holdings:
            symbol = holding['symbol']
            asset_class = holding['assetClass']
            current_value = float(holding['currentValue'])
            
            # Get historical performance
            asset_impact = self._calculate_asset_impact(
                symbol, 
                asset_class, 
                start_date, 
                end_date,
                current_value
            )
            
            affected_assets.append(asset_impact)
        
        # Calculate portfolio-level metrics
        final_value = sum(a['finalValue'] for a in affected_assets)
        total_return = ((final_value - initial_value) / initial_value) * 100
        
        # Generate timeline (60 points for smooth animation)
        timeline = self._generate_timeline(
            holdings, 
            start_date, 
            end_date, 
            initial_value
        )
        
        # Calculate max drawdown from timeline
        portfolio_values = [point['portfolioValue'] for point in timeline]
        max_drawdown = self._calculate_max_drawdown(portfolio_values, initial_value)
        
        # Calculate volatility
        returns = [
            (portfolio_values[i] - portfolio_values[i - 1]) / portfolio_values[i - 1]
            for i in range(1, len(portfolio_values))
            if portfolio_values[i - 1] != 0
        ]
        volatility = float(np.std(returns) * np.sqrt(252) * 100) if returns else 0.0  # Annualized
        
        # Estimate recovery days (simplified)
        recovery_days = max(0, abs(int(total_return * 5)))
        
        return {
            'scenarioId': scenario_id,
            'totalReturn': total_return,
            'maxDrawdown': max_drawdown,
            'recoveryDays': recovery_days,
            'volatility': volatility,
            'affectedAssets': affected_assets,
            'timeline': timeline
        }
    
    def _calculate_asset_impact(self, symbol, asset_class, start_date, end_date, current_value):
        """Calculate impact on a single asset"""
        try:
            series, used_symbol, used_proxy = self._get_effective_series(
                symbol,
                asset_class,
                start_date,
                end_date
            )

            if series is None:
                return {
                    'symbol': symbol,
                    'name': symbol,
                    'initialValue': float(current_value),
                    'finalValue': float(current_value),
                    'percentChange': 0.0,
                    'contribution': 0.0
                }

            first_price = float(series.iloc[0])
            last_price = float(series.iloc[-1])
            if first_price == 0:
                return {
                    'symbol': symbol,
                    'name': symbol,
                    'initialValue': float(current_value),
                    'finalValue': float(current_value),
                    'percentChange': 0.0,
                    'contribution': 0.0
                }

            percent_change = ((last_price - first_price) / first_price) * 100

            # Add extra volatility only when crypto falls back to proxy.
            if used_proxy and asset_class == 'crypto':
                percent_change *= 1.5
            
            final_value = current_value * (1 + percent_change / 100)
            contribution = ((final_value - current_value) / current_value) * 100
            
            return {
                'symbol': symbol,
                'name': symbol,
                'initialValue': float(current_value),
                'finalValue': float(final_value),
                'percentChange': float(percent_change),
                'contribution': float(contribution),
                'sourceSymbol': used_symbol
            }
            
        except Exception as e:
            print(f"Error fetching {symbol}: {e}")
            return {
                'symbol': symbol,
                'name': symbol,
                'initialValue': float(current_value),
                'finalValue': float(current_value),
                'percentChange': 0.0,
                'contribution': 0.0
            }
    
    def _use_proxy_calculation(self, symbol, asset_class, start_date, end_date, current_value):
        """Use proxy ticker when asset data not available"""
        proxy = self.PROXY_TICKERS.get(asset_class, 'SPY')
        
        try:
            data = yf.download(
                proxy, 
                start=start_date, 
                end=end_date, 
                progress=False,
                auto_adjust=True
            )
            
            if not data.empty and len(data) >= 2:
                first_price = data['Close'].iloc[0]
                last_price = data['Close'].iloc[-1]
                percent_change = ((last_price - first_price) / first_price) * 100
                
                # Apply volatility multiplier for certain asset classes
                if asset_class == 'crypto':
                    percent_change *= 1.5  # Crypto is more volatile
                elif asset_class == 'bond':
                    percent_change *= 0.3  # Bonds are less volatile
                
                final_value = current_value * (1 + percent_change / 100)
                contribution = ((final_value - current_value) / current_value) * 100
                
                return {
                    'symbol': symbol,
                    'name': symbol,
                    'initialValue': current_value,
                    'finalValue': final_value,
                    'percentChange': percent_change,
                    'contribution': contribution
                }
        except Exception as e:
            print(f"Error with proxy {proxy}: {e}")
        
        # Ultimate fallback: assume no change
        return {
            'symbol': symbol,
            'name': symbol,
            'initialValue': current_value,
            'finalValue': current_value,
            'percentChange': 0.0,
            'contribution': 0.0
        }
    
    def _generate_timeline(self, holdings, start_date, end_date, initial_value):
        """Generate timeline with 60 points for smooth animation"""
        timeline = []

        # Build aligned value paths for each holding.
        all_paths = []
        for holding in holdings:
            symbol = holding['symbol']
            asset_class = holding['assetClass']

            try:
                series, _, used_proxy = self._get_effective_series(
                    symbol,
                    asset_class,
                    start_date,
                    end_date
                )

                if series is None or len(series) < 2:
                    continue

                start_price = float(series.iloc[0])
                if start_price == 0:
                    continue

                normalized = series / start_price
                value_path = normalized * float(holding['currentValue'])

                if used_proxy and asset_class == 'crypto':
                    # Increase movement around start value when crypto uses proxy path.
                    value_path = float(holding['currentValue']) + (value_path - float(holding['currentValue'])) * 1.5

                all_paths.append(value_path.rename(symbol))
            except Exception as e:
                print(f"Error in timeline for {symbol}: {e}")
                continue

        if not all_paths:
            # Fallback: linear interpolation
            for i in range(60):
                timeline.append({
                    'date': start_date,
                    'portfolioValue': initial_value,
                    'percentChange': 0.0,
                    'timestamp': int(datetime.now().timestamp() * 1000) + i * 50
                })
            return timeline

        combined = pd.concat(all_paths, axis=1, join='inner').dropna()
        if combined.empty:
            # Fallback
            for i in range(60):
                timeline.append({
                    'date': start_date,
                    'portfolioValue': initial_value,
                    'percentChange': 0.0,
                    'timestamp': int(datetime.now().timestamp() * 1000) + i * 50
                })
            return timeline

        portfolio_series = combined.sum(axis=1)

        # Sample 60 points evenly across the timeline
        indices = np.linspace(0, len(portfolio_series) - 1, min(60, len(portfolio_series)), dtype=int)

        for idx in indices:
            date = portfolio_series.index[idx]
            portfolio_value = float(portfolio_series.iloc[idx])
            percent_change = ((portfolio_value - initial_value) / initial_value) * 100

            timeline.append({
                'date': date.strftime('%Y-%m-%d'),
                'portfolioValue': portfolio_value,
                'percentChange': float(percent_change),
                'timestamp': int(date.timestamp() * 1000)
            })

        return timeline
    
    def _calculate_max_drawdown(self, values, initial_value):
        """Calculate maximum drawdown as percentage"""
        numeric_values = np.asarray(values, dtype=float)
        if numeric_values.size == 0:
            return 0.0
        min_value = float(np.min(numeric_values))
        max_drawdown = ((min_value - initial_value) / initial_value) * 100
        return float(max_drawdown)


# Singleton instance
scenario_service = ScenarioService()

File: finai-dashboard/src/features/scenario/services/scenarioService.ts
import apiClient from '../../../services/apiClient'
import type { 
	ScenarioImpact, 
	ScenarioRequest,
	PortfolioHolding 
} from '../types/scenario.types'

class ScenarioService {
	private readonly baseUrl = '/api/scenario'

	/**
	 * Calculate scenario impact on portfolio
	 * Backend will compute how portfolio would perform during historical event
	 * based on asset correlations and historical data
	 */
	async calculateScenario(request: ScenarioRequest): Promise<ScenarioImpact> {
		console.log('🎯 Scenario Lab: Requesting calculation from backend...', request)
		
		try {
			const response = await apiClient.post<{ success: boolean, data: ScenarioImpact }>(
				`${this.baseUrl}/calculate`,
				request
			)
			
			// Check if backend returned success
			if (response.data.success && response.data.data) {
				console.log('✅ Scenario Lab: Using REAL HISTORICAL DATA from yfinance backend')
				console.log('📊 Total Return:', response.data.data.totalReturn.toFixed(2) + '%')
				console.log('📉 Max Drawdown:', response.data.data.maxDrawdown.toFixed(2) + '%')
				return response.data.data
			}
			
			// Fallback to mock data if backend response is invalid
			console.warn('⚠️ Backend returned invalid response, using mock data')
			return this.getMockScenarioImpact(request)
			
		} catch (error) {
			console.error('❌ Error calculating scenario:', error)
			console.warn('⚠️ Backend unavailable - using mock data as fallback')
			console.warn('💡 Make sure Flask backend is running on port 5001')
			return this.getMockScenarioImpact(request)
		}
	}

	/**
	 * Get historical performance data for a specific asset during a scenario
	 */
	async getAssetHistoricalData(symbol: string, scenarioId: string) {
		try {
			const response = await apiClient.get(
				`${this.baseUrl}/asset/${symbol}/scenario/${scenarioId}`
			)
			return response.data
		} catch (error) {
			console.error('Error fetching asset historical data:', error)
			throw error
		}
	}

	/**
	 * Mock data generator for frontend development
	 * This simulates what the backend will return
	 */
	private getMockScenarioImpact(request: ScenarioRequest): ScenarioImpact {
		const { scenarioId, holdings } = request
		const startValue = holdings.reduce((sum, h) => sum + h.currentValue, 0)
		
		// Different scenarios have different impacts based on asset classes
		const scenarioMultipliers: Record<string, Record<string, number>> = {
			'2008-financial-crisis': {
				stock: -0.57,
				etf: -0.55,
				bond: 0.05,
				crypto: -0.90, // Bitcoin wasn't really around
				commodity: -0.45
			},
			'covid-2020-crash': {
				stock: -0.34,
				etf: -0.33,
				bond: 0.08,
				crypto: -0.50, // Bitcoin actually recovered fast
				commodity: -0.25
			},
			'covid-recovery-2020': {
				stock: 1.20,
				etf: 1.15,
				bond: 0.03,
				crypto: 4.50, // Bitcoin soared
				commodity: 0.40
			},
			'crypto-winter-2022': {
				stock: -0.18,
				etf: -0.18,
				bond: -0.13,
				crypto: -0.77,
				commodity: 0.15
			},
			'post-2008-recovery': {
				stock: 4.00,
				etf: 3.85,
				bond: 0.45,
				crypto: 100000.00, // Bitcoin went from pennies to thousands
				commodity: 0.20
			},
			'dot-com-bubble': {
				stock: -0.49,
				etf: -0.45,
				bond: 0.20,
				crypto: 0,
				commodity: -0.10
			},
			'black-monday-1987': {
				stock: -0.226,
				etf: -0.226,
				bond: 0.02,
				crypto: 0,
				commodity: -0.15
			},
			'2022-bear-market': {
				stock: -0.25,
				etf: -0.24,
				bond: -0.14,
				crypto: -0.64,
				commodity: 0.25
			}
		}

		const multipliers = scenarioMultipliers[scenarioId] || {
			stock: -0.20,
			etf: -0.20,
			bond: 0,
			crypto: -0.30,
			commodity: -0.10
		}

		// Calculate impact per asset
		const affectedAssets = holdings.map(holding => {
			const multiplier = multipliers[holding.assetClass] || -0.20
			const finalValue = holding.currentValue * (1 + multiplier)
			const percentChange = multiplier * 100

			return {
				symbol: holding.symbol,
				name: holding.symbol, // In real impl, fetch full name
				initialValue: holding.currentValue,
				finalValue,
				percentChange,
				contribution: ((finalValue - holding.currentValue) / startValue) * 100
			}
		})

		const endValue = affectedAssets.reduce((sum, a) => sum + a.finalValue, 0)
		const totalReturn = ((endValue - startValue) / startValue) * 100

		// Generate timeline with realistic animation points
		const timeline = this.generateTimeline(startValue, endValue, scenarioId)

		// Calculate max drawdown
		const minValue = Math.min(...timeline.map(p => p.portfolioValue))
		const maxDrawdown = ((minValue - startValue) / startValue) * 100

		return {
			scenarioId,
			totalReturn,
			maxDrawdown,
			recoveryDays: Math.abs(Math.round(totalReturn * 5)), // Mock recovery time
			volatility: Math.abs(totalReturn * 0.5), // Mock volatility
			affectedAssets,
			timeline
		}
	}

	/**
	 * Generate realistic timeline for animation
	 * Creates data points that show gradual change with acceleration/deceleration
	 */
	private generateTimeline(startValue: number, endValue: number, scenarioId: string): any[] {
		const points = 60 // 60 frames for smooth animation
		const timeline = []
		const change = endValue - startValue
		const isCrash = change < 0

		for (let i = 0; i <= points; i++) {
			const t = i / points
			
			// Different easing based on scenario type
			let easedT: number
			if (isCrash) {
				// Crashes accelerate (fast drop, then plateau)
				if (scenarioId === 'black-monday-1987') {
					// Very sudden drop
					easedT = Math.pow(t, 0.5)
				} else {
					// Gradual acceleration
					easedT = t * t
				}
			} else {
				// Bull runs are more gradual
				easedT = 1 - Math.pow(1 - t, 3)
			}

			// Add some noise for realism
			const noise = (Math.random() - 0.5) * (Math.abs(change) * 0.02)
			const value = startValue + (change * easedT) + noise
			const percentChange = ((value - startValue) / startValue) * 100

			timeline.push({
				date: new Date(Date.now() + i * 50).toISOString(),
				portfolioValue: value,
				percentChange,
				timestamp: Date.now() + i * 50
			})
		}

		return timeline
	}

	/**
	 * Convert current portfolio to holdings format for scenario calculation
	 */
	convertPortfolioToHoldings(portfolio: any): PortfolioHolding[] {
		// This will be implemented based on actual portfolio structure
		// For now, return mock structure
		if (!portfolio?.accounts) return []

		const holdings: PortfolioHolding[] = []
		
		portfolio.accounts.forEach((account: any) => {
			account.positions?.forEach((position: any) => {
				holdings.push({
					symbol: position.symbol || position.ticker,
					quantity: position.quantity || position.shares,
					currentValue: position.marketValue || position.value,
					assetClass: this.determineAssetClass(position.symbol)
				})
			})
		})

		return holdings
	}

	/**
	 * Determine asset class from symbol
	 * In production, this should be fetched from backend/database
	 */
	private determineAssetClass(symbol: string): 'stock' | 'crypto' | 'bond' | 'etf' | 'commodity' {
		const crypto = ['BTC', 'ETH', 'BITCOIN', 'ETHEREUM', 'BNB', 'SOL', 'ADA', 'DOT']
		const etf = ['SPY', 'VOO', 'VFV', 'VTI', 'QQQ', 'IVV', 'VEA', 'IEFA', 'AGG', 'BND']
		const bond = ['TLT', 'IEF', 'SHY', 'LQD', 'HYG', 'MUB', 'BOND']
		const commodity = ['GLD', 'SLV', 'USO', 'UNG', 'DBA', 'GOLD', 'SILVER']

		const upperSymbol = symbol.toUpperCase()
		
		if (crypto.some(c => upperSymbol.includes(c))) return 'crypto'
		if (etf.some(e => upperSymbol === e)) return 'etf'
		if (bond.some(b => upperSymbol === b)) return 'bond'
		if (commodity.some(c => upperSymbol === c)) return 'commodity'
		
		return 'stock'
	}
}

export default new ScenarioService()
